consul/acl_replication.go

```diff
@@ -0,0 +1,351 @@
+			changes = append(changes, &structs.ACLRequest{
+				Op:  structs.ACLSet,
+				ACL: *(remoteIter.Front()),
+			})
+			remoteIter.Next()
+			continue
+		}
+
+		// If local has something remote doesn't, we delete it.
+		if localIter.Front().ID < remoteIter.Front().ID {
+			changes = append(changes, &structs.ACLRequest{
+				Op:  structs.ACLDelete,
+				ACL: *(localIter.Front()),
+			})
+			localIter.Next()
+			continue
+		}
+
+		// Local and remote have an ACL with the same ID, so we might
+		// need to compare them.
+		l, r := localIter.Front(), remoteIter.Front()
+		if r.RaftIndex.ModifyIndex > lastRemoteIndex && !r.IsSame(l) {
+			changes = append(changes, &structs.ACLRequest{
+				Op:  structs.ACLSet,
+				ACL: *r,
+			})
+		}
+		localIter.Next()
+		remoteIter.Next()
+	}
+	return changes
+}
+
+// FetchLocalACLs returns the ACLs in the local state store.
+func (s *Server) fetchLocalACLs() (structs.ACLs, error) {
+	_, local, err := s.fsm.State().ACLList()
+	if err != nil {
+		return nil, err
+	}
+	return local, nil
+}
+
+// FetchRemoteACLs is used to get the remote set of ACLs from the ACL
+// datacenter. The lastIndex parameter is a hint about which remote index we
+// have replicated to, so this is expected to block until something changes.
+func (s *Server) fetchRemoteACLs(lastRemoteIndex uint64) (*structs.IndexedACLs, error) {
+	args := structs.DCSpecificRequest{
+		Datacenter: s.config.ACLDatacenter,
+		QueryOptions: structs.QueryOptions{
+			Token:         s.config.ACLReplicationToken,
+			MinQueryIndex: lastRemoteIndex,
+			AllowStale:    true,
+		},
+	}
+	var remote structs.IndexedACLs
+	if err := s.RPC("ACL.List", &args, &remote); err != nil {
+		return nil, err
+	}
+	return &remote, nil
+}
+
+// UpdateLocalACLs is given a list of changes to apply in order to bring the
+// local ACLs in-line with the remote ACLs from the ACL datacenter.
+func (s *Server) updateLocalACLs(changes structs.ACLRequests) error {
+	var ops int
+	start := time.Now()
+	for _, change := range changes {
+		// Do a very simple rate limit algorithm where we check every N
+		// operations and wait out to the second before we continue. If
+		// it's going slower than that, the sleep time will be negative
+		// so we will just keep going without delay.
+		if ops >= s.config.ACLReplicationApplyLimit {
+			elapsed := time.Now().Sub(start)
+			time.Sleep(1*time.Second - elapsed)
+			ops, start = 0, time.Now()
+		}
+		ops++
+
+		// Note that we are using the single ACL interface here and not
+		// performing all this inside a single transaction. This is OK
+		// for two reasons. First, there's nothing else other than this
+		// replication routine that alters the local ACLs, so there's
+		// nothing to contend with locally. Second, if an apply fails
+		// in the middle (most likely due to losing leadership), the
+		// next replication pass will clean up and check everything
+		// again.
+		var reply string
+		if err := aclApplyInternal(s, change, &reply); err != nil {
```

Comment:
I kind of went back and forth on that. Using the existing plumbing saved some work, and I think in normal steady-state usage you'll take an initial hit and then trickle single ACLs in over time, so it didn't seem worth the extra effort. Also the writes are rate limited so this can't go too fast. If you think this is too jank I can add a batch thing though :-) If we added a batch I was going to add it to the Txn interface but not expose it externally on /v1/txn, just use it internally over the RPC (for now).